text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+reduced, making them a more viable pattern.
+
+### Namespaced Arguments
+
+In legacy components, arguments are set as properties directly on the class, and
+_mutable_. This leads to some confusing behavior, because the named argument
+syntax in templates refers to an _immutable_ copy of argument values. For
+instance, given this component:
+
+```js
+// components/welcome.js
+export default Component.extend({
+  firstName: 'Jen',
+  lastName: 'Weber'
+});
+```
+```hbs
+<!-- templates/components/welcome.hbs  -->
+Hey there, {{@firstName}} {{@lastName}}!
+```
+
+And this invocation:
+
+```hbs
+<Welcome />
+```
+
+You might expect that the result would be:
+
+```hbs
+Hey there, Jen Weber!
+```
+
+However, `{{@firstName}}` and `{{@lastName}}` would actually be empty values.
+They refer _directly_ to the arguments passed into the _invocation_ of the
+component, and to get the results we wanted, we would have to invoke it like so:
+
+```hbs
+<Welcome @firstName="Jen" @lastName="Weber" />
+```
+
+The advantage in this is that named arguments are fully transparent. When seen
+in a template, users can know without a doubt that the named argument was a
+value passed in from the invocation. Likewise, when they see a standard binding
+to a value like `{{this.firstName}}`, they know this is a value defined on the
+component - it could a computed property, it could come from a service, it could
+be a random value, but it is _not_ an argument.
+
+Glimmer components align their argument access with named args by making
+arguments available exclusively on a frozen object, `this.args`. Attempting to
+modify `this.args` or will hard error, meaning that like templates, users will
+always be able to refer to `this.args` as the canonical state of the values
+passed to the component invocation.
+
+Immutable arguments make reasoning about the state of a component simpler ("Was
+that a user provided value or a default/mutated/computed value?" becomes "Was it
+an argument or not?"), and encourages use of `{{@arg}}` syntax in templates
+where appropriate. At scale, this makes reading templates even easier, since
+more information is encoded in the template itself.
+
+### Minimal Classes
+
+Legacy components are _large_ classes, with lots of built up functionality and
+debt from over the years. The total list of default properties and hooks
+(including inherited ones) includes:
+
+* **13** Standard lifecycle hooks, such as
+  `didInsertElement`/`willDestroyElement` and `didUpdate`.
+* **29** Event handlers, such as `click`, `mouseEnter`, and `dragStart`.
+* **9** element/element customization properties, such as `element` and
+  `tagName`.
+* **21** standard framework functions, such as `get`/`set`,
+  `addObserver`/`removeObserver` and `toggleProperty`.
+
+Coming from a class hierarchy that is 4 levels deep (`Component` -> `CoreView`
+-> `EmberObject` -> `CoreObject`, with a number mixins included along the way).
```

Comment:
It may be worth saying that Ember.Component has like 10-20 mixins or so. (Checking emberaddons.com, it looks like its 19.....)